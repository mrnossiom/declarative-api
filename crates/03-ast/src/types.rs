use lexer::{span::Span, symbols::Symbol};

#[derive(Debug)]
pub struct Api {
	pub attrs: Vec<Attribute>,
	pub meta: Metadata,
	pub items: Vec<Item>,
	pub span: Span,
}

#[derive(Debug)]
pub struct Metadata {
	pub name: String,
	pub description: String,
	pub licence: String,
	pub urls: Vec<String>,
}

#[derive(Debug)]
pub struct Item {
	pub attrs: Vec<Attribute>,
	pub kind: ItemKind,
	pub span: Span,
	pub ident: Ident,
}

#[derive(Debug)]
pub enum ItemKind {
	Meta(Meta),

	Scope(ScopeKind),
	Path(Path),
	Model(Model),
	Headers(Headers),
	Verb(Verb),
	StatusCode(StatusCode),
}

#[derive(Debug)]
pub struct Verb {
	pub method: String,
	pub items: Vec<Item>,
}

#[derive(Debug)]
pub struct StatusCode {
	pub code: u16,
	pub items: Vec<Item>,
}

#[derive(Debug)]
pub struct Meta {
	pub fields: Vec<MetaField>,
}

#[derive(Debug)]
pub struct Headers {
	pub headers: Vec<HeaderField>,
}

#[derive(Debug)]
pub struct HeaderField {
	pub ident: Ident,
	pub attrs: Vec<Attribute>,
	pub span: Span,
}

#[derive(Debug)]
pub struct Path {
	pub path: PathKind,
	pub items: Vec<Item>,
}

#[derive(Debug)]
pub enum PathKind {
	String(Ident),
	Variable(Ident),
	Complex(Vec<Self>),
}

#[derive(Debug)]
pub struct Model {
	pub fields: Vec<ModelField>,
}
#[derive(Debug)]
pub struct ModelField {
	pub ident: Ident,
	pub ty: Type,
	pub attrs: Vec<Attribute>,
	pub span: Span,
}

#[derive(Debug)]
pub struct MetaField {
	pub ident: Ident,
	pub value: MetaFieldKind,
	pub span: Span,
}

#[derive(Debug)]
pub enum MetaFieldKind {
	Str(String),
	Bool(bool),
	Int(u64),
	Float(f64),
	Vec(Vec<Self>),
}

#[derive(Debug)]
pub struct Type(pub(crate) String);

#[derive(Debug)]
pub struct Attribute {
	pub ident: Ident,
	pub value: String,
	pub span: Span,
}

#[derive(Debug)]
pub enum ScopeKind {
	Loaded {
		items: Vec<Item>,
		/// Whether the scope was defined inline or in an external file.
		inline: bool,
		span: Span,
	},
	Unloaded,
}

#[derive(Debug)]
pub struct Ident {
	pub symbol: Symbol,
	pub span: Span,
}
