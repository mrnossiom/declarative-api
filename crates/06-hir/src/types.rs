#[derive(Debug)]
pub struct Version {
	pub major: u32,
	pub minor: u32,
	pub patch: u32,
}

#[derive(Debug)]
pub struct ApiMetadata {
	pub name: Option<String>,
	pub version: Option<Version>,
	pub licence: Option<String>,
	pub urls: Vec<String>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Response {
	pub status_code: u32,
	pub body: Vec<KeyValuePair>,
	pub headers: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Method {
	pub method: http::Method,
	pub responses: Vec<Response>,
	pub headers: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Path {
	pub name: String,
	pub scopes: Vec<Scope>,
	pub paths: Vec<Path>,
	pub methods: Vec<Method>,
	pub headers: Vec<KeyValuePair>,
	pub parameters: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub metadata: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

pub type KeyValuePairParameter = (String, String);

#[derive(Debug, Clone)]
pub struct KeyValuePair {
	pub key: String,
	pub type_: Type,
	pub description: String,
	pub parameters: Vec<KeyValuePairParameter>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Scope {
	pub name: String,
	pub scopes: Vec<Scope>,
	pub models: Vec<Model>,
	pub paths: Vec<Path>,
	pub methods: Vec<Method>,
	pub headers: Vec<KeyValuePair>,
	pub parameters: Vec<KeyValuePair>,
	pub query: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct Model {
	pub name: String,
	pub model_body: Vec<KeyValuePair>,
	pub comment: Option<String>,
}

#[derive(Debug)]
pub struct ApiData {
	pub scopes: Vec<Scope>,
}

#[derive(Debug)]
pub struct Api {
	pub metadata: ApiMetadata,
	pub data: ApiData,
}

#[derive(Debug, Clone)]
pub enum Type {
	Int,

	String,

	Model(String),
	Object(Vec<KeyValuePair>),
	List(Vec<Type>),
}
