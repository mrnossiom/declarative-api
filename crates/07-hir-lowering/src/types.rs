use std::{
	cmp::Ordering,
	collections::{BTreeMap, HashMap, HashSet},
	fmt,
	hash::Hash,
	path::Path,
};

#[derive(Debug, Clone)]
pub struct Version {
	pub major: u32,
	pub minor: u32,
	pub patch: u32,
}

impl fmt::Display for Version {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
	}
}

impl PartialOrd for Version {
	// Required method
	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
		self.major
			.partial_cmp(&other.major)
			.and_then(|ord| match ord {
				Ordering::Equal => self.minor.partial_cmp(&other.minor),
				_ => Some(ord),
			})
			.and_then(|ord| match ord {
				Ordering::Equal => self.patch.partial_cmp(&other.patch),
				_ => Some(ord),
			})
	}
}

impl PartialEq for Version {
	fn eq(&self, other: &Self) -> bool {
		self.major == other.major && self.minor == other.minor && self.patch == other.patch
	}
}

pub trait ResolveModels {
	fn resolve_models(
		&mut self,
		scope: &ScopePath,
		intermediate_representation: IntermediateRepresentation,
	);
}

pub type ScopePath = Vec<String>;

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub struct KeyValuePairValue {
	pub type_: Type,
	pub description: String,
	pub parameters: KeyValuePairValueParameters,
	pub comment: Option<String>,
}

impl KeyValuePairValue {
	#[must_use]
	pub fn from_ast_key_value_pair_ref(
		key_value_pair: &ast::types::KeyValuePair,
	) -> (String, Self) {
		(
			key_value_pair.key.clone(),
			Self {
				type_: Type::from_ast_type(key_value_pair.type_.clone()),
				description: key_value_pair.description.clone(),
				parameters: Self::parse_parameters(&key_value_pair.parameters),
				comment: key_value_pair.comment.clone(),
			},
		)
	}
	#[must_use]
	#[allow(clippy::missing_panics_doc)]
	pub fn map_from_ast_key_value_pair_vec(
		key_value_pair_vec: &[ast::types::KeyValuePair],
	) -> BTreeMap<String, Self> {
		let mut btreemap = BTreeMap::new();
		for field in key_value_pair_vec {
			let param_btreemap = Self::parse_parameters(&field.parameters);
			assert!(
				!btreemap.contains_key(&field.key),
				"a key-value pair has the same key defined multiple times"
			);
			btreemap.insert(
				field.key.clone(),
				Self {
					type_: Type::from_ast_type(field.type_.clone()),
					description: field.description.clone(),
					parameters: param_btreemap,
					comment: field.comment.clone(),
				},
			);
		}
		btreemap
	}
	// This merges the b BTreeMap into the a BTreeMap. In case of the same keys being defined in both BTreeMap, b's values has priority.
	pub fn merge(a: &mut BTreeMap<String, Self>, b: &BTreeMap<String, Self>) {
		for (key, value) in b {
			a.insert(key.clone(), value.clone());
		}
	}
	#[must_use]
	#[allow(clippy::missing_panics_doc)]
	pub fn parse_parameters(
		parameters: &[ast::types::KeyValuePairParameter],
	) -> BTreeMap<String, String> {
		let mut param_btreemap = BTreeMap::new();
		for parameter in parameters {
			assert!(
				!param_btreemap.contains_key(&parameter.0.clone()),
				"a key-value pair has the same parameter defined multiple times"
			);
			param_btreemap.insert(parameter.0.clone(), parameter.1.clone());
		}
		param_btreemap
	}
}

impl ResolveModels for BTreeMap<String, KeyValuePairValue> {
	fn resolve_models(
		&mut self,
		scope: &ScopePath,
		intermediate_representation: IntermediateRepresentation,
	) {
		for value in self.values_mut() {
			match &mut value.type_ {
				Type::Model(model_name) => {
					let mut found = false;
					let mut current_scope = scope.clone();
					while !found {
						found = intermediate_representation
							.scopes
							.get(&current_scope)
							.expect("missing scope")
							.models
							.contains_key(model_name);
						if !found {
							assert!(
								!current_scope.is_empty(),
								"unknown reference to model \"{model_name}\""
							);
							current_scope.truncate(current_scope.len() - 1);
						}
					}
					current_scope.push(model_name.clone());
					value.type_ = Type::ResolvedModel(current_scope);
				}
				Type::Object(child_object) => {
					child_object.resolve_models(scope, intermediate_representation.clone());
				}
				_ => (),
			};
		}
	}
}

#[derive(Debug, PartialEq, Clone, Eq, Hash)]
pub struct Method {
	pub parameters: BTreeMap<String, KeyValuePairValue>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub query_params: BTreeMap<String, KeyValuePairValue>,
	pub responses: BTreeMap<u32, Response>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub struct Response {
	pub body: BTreeMap<String, KeyValuePairValue>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Eq, Hash, Default)]
pub struct Endpoint {
	pub methods: BTreeMap<String, Method>,
}

#[derive(Debug, PartialEq, Eq, Clone, Default)]
pub struct Scope {
	pub child_scopes: HashSet<String>,
	pub models: BTreeMap<String, Model>,
	pub endpoints: HashMap<String, Endpoint>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone, Eq, Hash, Default)]
pub struct Model {
	pub model_body: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

pub type KeyValuePairValueParameters = BTreeMap<String, String>;

#[derive(Debug, PartialEq, Clone, Hash, Eq)]
pub enum Type {
	Int,

	String,

	Model(String),
	ResolvedModel(ScopePath),
	Object(BTreeMap<String, KeyValuePairValue>),
	List(Vec<Type>),
}

impl Type {
	#[must_use]
	pub fn from_ast_type(input: ast::types::Type) -> Self {
		match input {
			ast::types::Type::Int => Self::Int,
			ast::types::Type::String => Self::String,
			ast::types::Type::Model(model_name) => Self::Model(model_name),
			ast::types::Type::Object(fields) => {
				Self::Object(KeyValuePairValue::map_from_ast_key_value_pair_vec(&fields))
			}
			ast::types::Type::List(list_obj) => {
				let mut out_list_obj = vec![];
				for obj in &list_obj {
					out_list_obj.push(Self::from_ast_type(obj.clone()));
				}
				Self::List(out_list_obj)
			}
		}
	}
}

#[derive(Debug, PartialEq, Clone)]
pub struct ApiMetadata {
	pub name: String,
	pub version: Version,
	pub urls: Vec<String>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq, Clone)]
pub struct IntermediateRepresentation {
	pub metadata: ApiMetadata,
	pub scopes: BTreeMap<ScopePath, Scope>,
}

impl Default for IntermediateRepresentation {
	fn default() -> Self {
		let mut scopes_map = BTreeMap::new();
		scopes_map.insert(vec![], Scope::default());
		Self {
			metadata: ApiMetadata {
				name: String::new(),
				version: Version {
					major: 0,
					minor: 0,
					patch: 0,
				},
				urls: vec![],
				comment: None,
			},
			scopes: scopes_map,
		}
	}
}

#[derive(Debug, Clone)]
pub struct ParserVariables {
	pub current_method: String,
	pub scope_path: ScopePath,
	pub endpoint_path: Box<Path>,
	pub headers: BTreeMap<String, KeyValuePairValue>,
	pub parameters: BTreeMap<String, KeyValuePairValue>,
	pub query_params: BTreeMap<String, KeyValuePairValue>,
	pub comment: Option<String>,
}

impl Default for ParserVariables {
	fn default() -> Self {
		Self {
			current_method: String::new(),
			scope_path: ScopePath::new(),
			endpoint_path: Path::new("/").into(),
			headers: BTreeMap::new(),
			parameters: BTreeMap::new(),
			query_params: BTreeMap::new(),
			comment: None,
		}
	}
}
