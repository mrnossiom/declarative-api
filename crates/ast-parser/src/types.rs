use std::{cmp::Ordering, collections::HashMap, fmt, vec};

#[derive(Debug, Eq, Hash, PartialEq)]
pub struct ScopePath {
	path: Vec<String>,
}

#[derive(Debug)]
pub struct Version {
	pub major: u32,
	pub minor: u32,
	pub patch: u32,
}

impl fmt::Display for Version {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
	}
}

impl PartialOrd for Version {
	// Required method
	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
		self.major
			.partial_cmp(&other.major)
			.and_then(|ord| match ord {
				Ordering::Equal => self.minor.partial_cmp(&other.minor),
				_ => Some(ord),
			})
			.and_then(|ord| match ord {
				Ordering::Equal => self.patch.partial_cmp(&other.patch),
				_ => Some(ord),
			})
	}
}

impl PartialEq for Version {
	fn eq(&self, other: &Self) -> bool {
		self.major == other.major && self.minor == other.minor && self.patch == other.patch
	}
}

#[derive(Debug, PartialEq)]
pub struct KeyValuePairValue {
	pub type_: Type,
	pub description: String,
	pub parameters: HashMap<String, String>,
	pub comment: String,
}

#[derive(Debug, PartialEq)]
pub struct Method {
	pub responses: Vec<Response>,
	pub headers: HashMap<String, KeyValuePairValue>,
	pub parameters: HashMap<String, KeyValuePairValue>,
	pub comment: String,
}

#[derive(Debug, PartialEq)]
pub struct Path {
	pub methods: Vec<Method>,
	pub headers: HashMap<String, KeyValuePairValue>,
	pub parameters: HashMap<String, KeyValuePairValue>,
	pub metadata: HashMap<String, KeyValuePairValue>,
	pub comment: String,
}

#[derive(Debug, PartialEq)]
pub struct Response {
	pub status_code: u32,
	pub body: HashMap<String, KeyValuePairValue>,
	pub headers: HashMap<String, KeyValuePairValue>,
	pub comment: String,
}

#[derive(Debug, PartialEq)]
pub struct Scope {
	pub child_scopes: Vec<Scope>,
	pub child_models: Vec<Model>,
	pub child_paths: Vec<Path>,
	pub methods: Vec<Method>,
	pub comment: String,
}

#[derive(Debug, PartialEq)]
pub struct Model {
	pub name: String,
	pub model_body: HashMap<String, KeyValuePairValue>,
	pub comment: String,
}

#[derive(Debug, PartialEq)]
pub enum Type {
	Int,

	String,

	Model(String),
	Object(Box<(String, KeyValuePairValue)>),
}

#[derive(Debug, PartialEq)]
pub struct ApiMetadata {
	pub name: Option<String>,
	pub version: Option<Version>,
	pub urls: Vec<String>,
	pub comment: Option<String>,
}

#[derive(Debug, PartialEq)]
pub struct IntermediateRepresentation {
	pub metadata: ApiMetadata,
	pub scopes: HashMap<ScopePath, Scope>,
}

impl IntermediateRepresentation {
	pub fn new() -> IntermediateRepresentation {
		IntermediateRepresentation {
			metadata: ApiMetadata {
				name: None,
				version: None,
				urls: vec![],
				comment: None,
			},
			scopes: HashMap::new(),
		}
	}
}

#[derive(Debug)]
pub struct ParserVariables {
	pub scope_path: ScopePath,
	pub endpoint_path: String,
	pub headers: HashMap<String, KeyValuePairValue>,
	pub parameters: HashMap<String, KeyValuePairValue>,
	pub comment: String,
}

impl ParserVariables {
	pub fn new() -> ParserVariables {
		ParserVariables {
			scope_path: ScopePath { path: vec![] },
			endpoint_path: "".into(),
			headers: HashMap::new(),
			parameters: HashMap::new(),
			comment: "".into(),
		}
	}
}
