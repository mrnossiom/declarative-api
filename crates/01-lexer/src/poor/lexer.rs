use super::{
	cursor::Cursor,
	token::{
		DocStyle,
		LiteralKind::*,
		Token,
		TokenKind::{self, *},
	},
};
use unicode_xid::UnicodeXID;

impl<'a> Cursor<'a> {
	pub(crate) fn advance_token(&mut self) -> Token {
		let Some(first_char) = self.bump() else {
			return Token::new(Eof, 0);
		};

		let token_kind = match first_char {
			'#' => match self.first() {
				'#' | ' ' => self.line_comment(),
				_ => Pound,
			},

			// Whitespace sequence.
			c if is_whitespace(c) => self.whitespace(),

			// Identifier (this should be checked after other variant that can
			// start as identifier).
			c if is_id_start(c) => self.ident(),

			// Numeric literal.
			c @ '0'..='9' => self.number(c),

			// One-symbol tokens.
			';' => Semi,
			',' => Comma,
			'.' => Dot,
			'(' => OpenParenthesis,
			'{' => OpenBrace,
			'[' => OpenBracket,
			')' => CloseParenthesis,
			'}' => CloseBrace,
			']' => CloseBracket,
			'@' => At,
			'~' => Tilde,
			'?' => Question,
			':' => Colon,
			'$' => Dollar,
			'=' => Eq,
			'!' => Bang,
			'<' => Lt,
			'>' => Gt,
			'-' => Minus,
			'&' => And,
			'|' => Or,
			'+' => Plus,
			'*' => Star,
			'^' => Caret,
			'%' => Percent,

			// String literal.
			'"' => {
				let terminated = self.double_quoted_string();
				Literal(Str { terminated })
			}

			_ => Unknown,
		};

		let res = Token::new(token_kind, self.pos_within_token());
		self.reset_pos_within_token();
		res
	}

	fn line_comment(&mut self) -> TokenKind {
		debug_assert!(self.prev() == '#' && (self.first() == '#' || self.first() == ' '));
		self.bump();

		let style = match self.first() {
			'!' => Some(DocStyle::Inner),
			' ' => Some(DocStyle::Outer),
			_ => None,
		};

		self.eat_while(|c| c != '\n');
		LineComment(style)
	}

	fn whitespace(&mut self) -> TokenKind {
		debug_assert!(is_whitespace(self.prev()));
		self.eat_while(is_whitespace);
		Whitespace
	}

	fn ident(&mut self) -> TokenKind {
		debug_assert!(is_id_start(self.prev()));
		// Start is already eaten, eat the rest of identifier.
		self.eat_while(is_id_continue);
		// Known prefixes must have been handled earlier. So if
		// we see a prefix here, it is definitely an unknown prefix.
		match self.first() {
			c if !c.is_ascii() && unic_emoji_char::is_emoji(c) => self.fake_ident(),
			_ => Ident,
		}
	}

	fn fake_ident(&mut self) -> TokenKind {
		// Start is already eaten, eat the rest of identifier.
		self.eat_while(|c| {
			UnicodeXID::is_xid_continue(c)
				|| (!c.is_ascii() && unic_emoji_char::is_emoji(c))
				|| c == '\u{200d}'
		});

		InvalidIdent
	}

	/// Eats double-quoted string and returns true
	/// if string is terminated.
	fn double_quoted_string(&mut self) -> bool {
		debug_assert!(self.prev() == '"');
		while let Some(c) = self.bump() {
			match c {
				'"' => {
					return true;
				}
				'\\' if self.first() == '\\' || self.first() == '"' => {
					// Bump again to skip escaped character.
					self.bump();
				}
				_ => (),
			}
		}
		// End of file reached.
		false
	}

	fn number(&mut self, _c: char) -> TokenKind {
		self.eat_decimal_digits();

		Literal(Number)
	}
}

/// True if `c` is considered a whitespace
const fn is_whitespace(c: char) -> bool {
	// This is Pattern_White_Space.
	//
	// Note that this set is stable (ie, it doesn't change with different
	// Unicode versions), so it's ok to just hard-code the values.

	matches!(
		c,
		// Usual ASCII suspects
		'\u{0009}'   // \t
        | '\u{000A}' // \n
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{000D}' // \r
        | '\u{0020}' // space

        // NEXT LINE from latin1
        | '\u{0085}'

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK

        // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR
	)
}

/// True if `c` is valid as a first character of an identifier.
fn is_id_start(c: char) -> bool {
	// This is XID_Start OR '_' (which formally is not a XID_Start).
	c == '_' || UnicodeXID::is_xid_start(c)
}

/// True if `c` is valid as a non-first character of an identifier.
fn is_id_continue(c: char) -> bool {
	UnicodeXID::is_xid_continue(c)
}

/// The passed string is lexically an identifier.
fn _is_ident(string: &str) -> bool {
	let mut chars = string.chars();
	chars.next().map_or(false, |start| {
		is_id_start(start) && chars.all(is_id_continue)
	})
}
