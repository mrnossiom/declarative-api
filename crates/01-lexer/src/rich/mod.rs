mod enricher;
mod errors;
mod token;

pub use enricher::Enricher;
pub use token::*;
