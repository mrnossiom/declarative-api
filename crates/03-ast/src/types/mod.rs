use crate::P;

mod attr;
mod expr;
mod item;
mod node;
pub use attr::*;
pub use expr::*;
pub use item::*;
pub use node::*;
use session::Span;
use thin_vec::ThinVec;

#[derive(Debug, Clone)]
pub struct Api {
	pub attrs: AttrVec,
	pub items: ThinVec<P<Item>>,

	pub id: NodeId,
	pub span: Span,
}

// TODO: nope, not a String
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Type(pub String);
