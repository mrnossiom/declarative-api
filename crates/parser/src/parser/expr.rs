use crate::{error::UnexpectedToken, PResult, Parser};
use ast::types::{AttrStyle, Expr, ExprKind, FieldDef, NodeId, PropertyDef, P};
use lexer::rich::{Delimiter, LiteralKind, TokenKind};
use session::{
	sym,
	symbols::{attrs, kw},
	Ident,
};
use thin_vec::{thin_vec, ThinVec};
use tracing::instrument;

impl<'a> Parser<'a> {
	#[instrument(level = "DEBUG", skip(self))]
	pub(super) fn parse_expr(&mut self) -> PResult<P<Expr>> {
		let lo = self.token.span;

		let kind = self.parse_expr_kind()?;

		Ok(Self::make_expr(thin_vec![], kind, self.span(lo)))
	}

	#[instrument(level = "DEBUG", skip(self))]
	fn parse_expr_kind(&mut self) -> PResult<ExprKind> {
		if self.check(&TokenKind::OpenDelim(Delimiter::Bracket)) {
			self.parse_expr_array()
		} else {
			self.parse_expr_literal()
		}
	}

	#[instrument(level = "DEBUG", skip(self))]
	fn parse_expr_literal(&mut self) -> PResult<ExprKind> {
		if let TokenKind::Literal(kind, sym) = self.token.kind {
			self.bump();
			Ok(ExprKind::Literal(kind, sym))
		} else if let Some(Ident {
			symbol: sym @ (kw::True | kw::False),
			..
		}) = self.token.ident().map(Into::into)
		{
			self.bump();
			Ok(ExprKind::Literal(LiteralKind::Bool, sym))
		} else {
			// TODO: recover properly
			Err(UnexpectedToken {
				parsed: self.token.clone(),
				expected: TokenKind::Literal(LiteralKind::Str, sym!("remove")),
			}
			.into())
		}
	}

	#[instrument(level = "DEBUG", skip(self))]
	fn parse_expr_array(&mut self) -> PResult<ExprKind> {
		let mut items = ThinVec::default();

		self.expect(&TokenKind::OpenDelim(Delimiter::Bracket))?;
		while !self.eat(&TokenKind::CloseDelim(Delimiter::Bracket)) {
			items.push(self.parse_expr()?);
		}

		Ok(ExprKind::Array(items))
	}

	#[instrument(level = "DEBUG", skip(self))]
	pub(super) fn parse_property_defs(&mut self) -> PResult<ThinVec<P<PropertyDef>>> {
		let mut props = ThinVec::new();
		while let Some(prop) = self.parse_property_def()? {
			props.push(prop);
		}
		Ok(props)
	}

	#[instrument(level = "DEBUG", skip(self))]
	fn parse_property_def(&mut self) -> PResult<Option<P<PropertyDef>>> {
		let lo = self.token.span;

		let mut attrs = self.parse_outer_attrs()?;

		let Some(ident) = self.eat_ident() else {
			return Ok(None);
		};
		let value = self.parse_expr()?;

		if let Some(attrs_) = self.parse_inline_attrs()? {
			attrs.extend(attrs_);
		}

		Ok(Some(Self::make_property_def(
			attrs,
			ident,
			value,
			self.span(lo),
		)))
	}

	#[instrument(level = "DEBUG", skip(self))]
	pub(super) fn parse_field_defs(&mut self) -> PResult<ThinVec<P<FieldDef>>> {
		let mut fields = ThinVec::new();
		while let Some(field) = self.parse_field_def()? {
			fields.push(field);
		}
		Ok(fields)
	}

	#[instrument(level = "DEBUG", skip(self))]
	fn parse_field_def(&mut self) -> PResult<Option<P<FieldDef>>> {
		let lo = self.token.span;

		let mut attrs = self.parse_outer_attrs()?;

		let Some(ident) = self.eat_ident() else {
			return Ok(None);
		};
		let ty = self.parse_ty()?;

		// Extract the optional literal after the type and transform into sugar for a meta attr named `description`
		if let Ok(ExprKind::Literal(LiteralKind::Str, sym)) = self.parse_expr_literal() {
			attrs.push(Self::make_meta_attr(
				Ident::new(attrs::description, self.prev_token.span),
				Some(P(Expr {
					attrs: ThinVec::new(),
					kind: ExprKind::Literal(LiteralKind::Str, sym),
					id: NodeId::DUMMY,
					span: self.prev_token.span,
				})),
				AttrStyle::Inline,
				self.prev_token.span,
			));
		}

		if let Some(inline_attrs) = self.parse_inline_attrs()? {
			attrs.extend(inline_attrs);
		}

		Ok(Some(Self::make_field_def(attrs, ident, ty, self.span(lo))))
	}
}

#[cfg(test)]
mod tests {
	use crate::assert_tokenize;

	assert_tokenize!(
		parse_field_defs,
		r#"
			## # Safety
			## This is a comment
			## This is a second line of comment
			Authorization long_string "The API Key of the User of the User" |@prefix: "Api-Key"|
			# ^ ident     ^ type      ^ sugar for description attr          ^ prefix attr

			X-Model string "The Model of the User"
		"#
	);
}
