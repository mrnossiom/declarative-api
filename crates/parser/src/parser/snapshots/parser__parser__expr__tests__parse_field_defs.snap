---
source: crates/parser/src/parser/expr.rs
expression: fields
---
[
    FieldDef {
        attrs: [
            Attribute {
                kind: DocComment(
                    Symbol(29, " # Safety"),
                ),
                style: Outer,
                id: AttrId(
                    0,
                ),
                span: Span {
                    start: BytePos(
                        1,
                    ),
                    end: BytePos(
                        12,
                    ),
                },
            },
            Attribute {
                kind: DocComment(
                    Symbol(30, " This is a comment"),
                ),
                style: Outer,
                id: AttrId(
                    1,
                ),
                span: Span {
                    start: BytePos(
                        13,
                    ),
                    end: BytePos(
                        33,
                    ),
                },
            },
            Attribute {
                kind: DocComment(
                    Symbol(31, " This is a second line of comment"),
                ),
                style: Outer,
                id: AttrId(
                    2,
                ),
                span: Span {
                    start: BytePos(
                        34,
                    ),
                    end: BytePos(
                        69,
                    ),
                },
            },
            Attribute {
                kind: Meta(
                    MetaAttr {
                        ident: Ident {
                            symbol: Symbol(17, "description"),
                            span: Span {
                                start: BytePos(
                                    96,
                                ),
                                end: BytePos(
                                    133,
                                ),
                            },
                        },
                        expr: Some(
                            Expr {
                                attrs: [],
                                kind: Literal(
                                    Str,
                                    Symbol(34, "The API Key of the User of the User"),
                                ),
                                id: NodeId(DUMMY),
                                span: Span {
                                    start: BytePos(
                                        96,
                                    ),
                                    end: BytePos(
                                        133,
                                    ),
                                },
                            },
                        ),
                    },
                ),
                style: Inline,
                id: AttrId(
                    3,
                ),
                span: Span {
                    start: BytePos(
                        96,
                    ),
                    end: BytePos(
                        133,
                    ),
                },
            },
            Attribute {
                kind: Meta(
                    MetaAttr {
                        ident: Ident {
                            symbol: Symbol(35, "prefix"),
                            span: Span {
                                start: BytePos(
                                    136,
                                ),
                                end: BytePos(
                                    142,
                                ),
                            },
                        },
                        expr: Some(
                            Expr {
                                attrs: [],
                                kind: Literal(
                                    Str,
                                    Symbol(36, "Api-Key"),
                                ),
                                id: NodeId(DUMMY),
                                span: Span {
                                    start: BytePos(
                                        144,
                                    ),
                                    end: BytePos(
                                        153,
                                    ),
                                },
                            },
                        ),
                    },
                ),
                style: Inline,
                id: AttrId(
                    4,
                ),
                span: Span {
                    start: BytePos(
                        135,
                    ),
                    end: BytePos(
                        153,
                    ),
                },
            },
        ],
        ident: Ident {
            symbol: Symbol(32, "Authorization"),
            span: Span {
                start: BytePos(
                    70,
                ),
                end: BytePos(
                    83,
                ),
            },
        },
        ty: Ty {
            kind: Path(
                Path {
                    segments: [
                        PathSegment {
                            ident: Ident {
                                symbol: Symbol(33, "long_string"),
                                span: Span {
                                    start: BytePos(
                                        84,
                                    ),
                                    end: BytePos(
                                        95,
                                    ),
                                },
                            },
                            id: NodeId(DUMMY),
                        },
                    ],
                    span: Span {
                        start: BytePos(
                            96,
                        ),
                        end: BytePos(
                            133,
                        ),
                    },
                },
            ),
            id: NodeId(DUMMY),
            span: Span {
                start: BytePos(
                    84,
                ),
                end: BytePos(
                    95,
                ),
            },
        },
        id: NodeId(DUMMY),
        span: Span {
            start: BytePos(
                1,
            ),
            end: BytePos(
                154,
            ),
        },
    },
    FieldDef {
        attrs: [
            Attribute {
                kind: Meta(
                    MetaAttr {
                        ident: Ident {
                            symbol: Symbol(17, "description"),
                            span: Span {
                                start: BytePos(
                                    249,
                                ),
                                end: BytePos(
                                    272,
                                ),
                            },
                        },
                        expr: Some(
                            Expr {
                                attrs: [],
                                kind: Literal(
                                    Str,
                                    Symbol(39, "The Model of the User"),
                                ),
                                id: NodeId(DUMMY),
                                span: Span {
                                    start: BytePos(
                                        249,
                                    ),
                                    end: BytePos(
                                        272,
                                    ),
                                },
                            },
                        ),
                    },
                ),
                style: Inline,
                id: AttrId(
                    5,
                ),
                span: Span {
                    start: BytePos(
                        249,
                    ),
                    end: BytePos(
                        272,
                    ),
                },
            },
        ],
        ident: Ident {
            symbol: Symbol(37, "X-Model"),
            span: Span {
                start: BytePos(
                    234,
                ),
                end: BytePos(
                    241,
                ),
            },
        },
        ty: Ty {
            kind: Path(
                Path {
                    segments: [
                        PathSegment {
                            ident: Ident {
                                symbol: Symbol(38, "string"),
                                span: Span {
                                    start: BytePos(
                                        242,
                                    ),
                                    end: BytePos(
                                        248,
                                    ),
                                },
                            },
                            id: NodeId(DUMMY),
                        },
                    ],
                    span: Span {
                        start: BytePos(
                            249,
                        ),
                        end: BytePos(
                            272,
                        ),
                    },
                },
            ),
            id: NodeId(DUMMY),
            span: Span {
                start: BytePos(
                    242,
                ),
                end: BytePos(
                    248,
                ),
            },
        },
        id: NodeId(DUMMY),
        span: Span {
            start: BytePos(
                234,
            ),
            end: BytePos(
                272,
            ),
        },
    },
]
